### PR TITLE
Feature/ implement handler and service logic for updating trees

### PR DESCRIPTION
**Pull request by Mohammad96Assaf**
refs: 
#78 , #104

 ~~## Note~~
-  ~~Removed `geometry = ST_GeomFromText($10, 4326)` from the query to resolve an issue caused by an inconsistent parameter type in the `UPDATE` query~~

### CODE REVIEW

File: internal/entities/tree.go [hunk 1]
```diff
@@ -19,3 +19,13 @@ type Tree struct {
...
+
+type TreeUpdate struct {
+	TreeClusterID *int32
+	Readonly      bool
+	PlantingYear  int32
+	Species       string
+	Number        string
+	Latitude      float64
+	Longitude     float64
+}
```
Comment (choffmann) [inline internal/entities/tree.go:31]: Do we really need this?

File: internal/storage/postgres/queries/trees.sql [hunk 2]
```diff
@@ -36,8 +36,7 @@ UPDATE trees SET
   latitude = $7,
   longitude = $8,
   readonly = $9,
-  watering_status = $10,
-  geometry = ST_GeomFromText($10, 4326)
+  watering_status = $10
```
Comment (choffmann) [inline internal/storage/postgres/queries/trees.sql:39]: That's not the solution to delete `geometry`. It should be like this

```
watering_status = $10,
geometry = ST_GeomFromText($11, 4326)
```

File: internal/service/domain/tree/tree.go [hunk 3]
```diff
@@ -62,3 +66,41 @@ func (s *TreeService) Delete(ctx context.Context, id int32) error {
...
+	}
+	if tu.Species != "" {
+		fn = append(fn, tree.WithSpecies(tu.Species))
+	}
+	if tu.Number != "" {
+		fn = append(fn, tree.WithTreeNumber(tu.Number))
+	}
+	if tu.Latitude != 0 && tu.Longitude != 0 {
+		fn = append(fn, tree.WithLatitude(tu.Latitude), tree.WithLongitude(tu.Longitude))
+	}
```
Comment (choffmann) [inline internal/service/domain/tree/tree.go:86]: Yes, okay... another solution, and I think a much cleaner solution, is to use some kind of validator, like https://pkg.go.dev/github.com/go-playground/validator/v10, which I already have somewhere here in this project

File: internal/service/domain/tree/tree.go [hunk 4]
```diff
@@ -62,3 +66,41 @@ func (s *TreeService) Delete(ctx context.Context, id int32) error {
...
+		return nil, handleError(fmt.Errorf("tree with ID %d is readonly and cannot be updated", id))
+	}
+	fn := make([]entities.EntityFunc[entities.Tree], 0)
+	if tu.TreeClusterID != nil {
+		treeCluster, err := s.treeClusterRepo.GetByID(ctx, *tu.TreeClusterID)
+		if err != nil {
+			return nil, handleError(fmt.Errorf("failed to find TreeCluster with ID %d: %w", *tu.TreeClusterID, err))
+		}
+		fn = append(fn, tree.WithTreeCluster(treeCluster))
+	}
```
Comment (choffmann) [inline internal/service/domain/tree/tree.go:74]: When a tree gets updated, the old tree cluster coordinate needs to be updated as well

**Review by choffmann — CHANGES_REQUESTED** [review]

File: internal/service/domain/tree/tree.go (hunk 4 above)
Comment (choffmann) [reply internal/service/domain/tree/tree.go:74]: see #98

File: internal/service/domain/tree/tree.go (hunk 3 above)
Comment (Mohammad96Assaf) [reply internal/service/domain/tree/tree.go]: I noticed that you don't use any kind of validation in your implementation for updating or creating tree clusters. I want to ask if we should follow the same approach in the tree cluster section?

File: internal/service/domain/tree/tree.go (hunk 4 above)
Comment (Mohammad96Assaf) [reply internal/service/domain/tree/tree.go:136]: Not only the old tree cluster, but I think the new one should also be updated, right?

**Review by Mohammad96Assaf — COMMENTED** [review]
@choffmann please review the latest edition. I am also waiting for the new component to continue with the functionality for updating the tree cluster coordinates.

File: internal/entities/tree.go [hunk 5]
```diff
@@ -28,3 +28,12 @@ type TreeCreate struct {
...
 	Longitude     float64
 }
+
+type TreeUpdate struct {
+	TreeClusterID *int32  `validate:"omitempty,gt=0"`
+	PlantingYear  int32   `validate:"omitempty,gt=0"`
+	Species       string  `validate:"omitempty"`
+	Number        string  `validate:"omitempty"`
+	Latitude      float64 `validate:"not-zero"`
+	Longitude     float64 `validate:"not-zero"`
```
[the remark this answers is not in the document]
Comment (choffmann) [reply internal/entities/tree.go:38]: and lat and long can both have a zero value

File: internal/service/domain/tree/tree.go [hunk 6]
```diff
@@ -99,5 +105,48 @@ func (s *TreeService) Delete(ctx context.Context, id int32) error {
...
+	if err != nil {
+		return nil, handleError(err)
+	}
+	// TODO: If a new tree cluster has been provided, update the coordinates of both the old tree cluster and the new one.
+	return updatedTree, nil
+}
+func notZero(fl validator.FieldLevel) bool {
+	value := fl.Field().Float()
+	return value != 0
+}
```
Comment (choffmann) [inline internal/service/domain/tree/tree.go:152]: Do we really need a custome validator here?

**Review by choffmann — CHANGES_REQUESTED** [review]
:+1:

File: internal/entities/tree.go [hunk 7]
```diff
@@ -28,3 +28,12 @@ type TreeCreate struct {
 	Latitude      float64
 	Longitude     float64
 }
+
+type TreeUpdate struct {
+	TreeClusterID *int32  `validate:"omitempty,gt=0"`
```
[the remark this answers is not in the document]
Comment (Mohammad96Assaf) [reply internal/entities/tree.go:33]: it means that the field is optional and not required for validation. If the field is set to its zero value (the default value for that type, such as 0 for numbers, "" for strings, or nil for pointers), the validator will **skip** further validation (e.g., gt=0 for TreeClusterID) for that field. 


https://pkg.go.dev/github.com/go-playground/validator/v10#hdr-Omit_Empty

**Review by choffmann — APPROVED** [review]

**Pull request merged**